File: python/lexer.py
from tokens import Token, TokenType, KEYWORDS
from typing import List


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.start = 0
        self.curr = 0
        self.line = 1
        self.tokens: List[Token] = []

    def advance(self) -> str:
        if self.curr >= len(self.source):
            return '\0'
        ch = self.source[self.curr]
        self.curr = self.curr + 1
        return ch

    def add_token(self, token_type: TokenType):
        self.tokens.append(
            Token(token_type, self.source[self.start : self.curr], self.line)
        )

    def peek(self) -> str:
        if self.curr >= len(self.source):
            return '\0'
        return self.source[self.curr]

    def lookahead(self, n=1) -> str:
        if self.curr+n >= len(self.source):
            return "\0"
        return self.source[self.curr + n]

    def match(self, expected) -> bool:
        if self.curr >= len(self.source) or self.source[self.curr] != expected:
            return False
        self.curr += 1
        return True

    def handle_number(self):
        while self.peek().isdigit():
            self.advance()
        if self.peek() == '.' and self.lookahead().isdigit():
            self.advance()
            while self.peek().isdigit():
                self.advance()
            self.add_token(TokenType.TOK_FLOAT)
        else:
            self.add_token(TokenType.TOK_INTEGER)

    def handle_identifier(self):
        while self.peek().isalnum() or self.peek() == '_':
            self.advance()
        self.add_token(KEYWORDS[self.source[self.start:self.curr]])

    def handle_string(self, ch: str):
        while self.peek() != ch and self.curr < len(self.source):
            self.advance()
        if self.curr >= len(self.source):
            raise SystemError(f"Unclosed quote on line {self.line}")
        self.advance()
        self.add_token(TokenType.TOK_STRING)


    def tokenize(self):
        while self.curr < len(self.source):
            self.start = self.curr
            ch = self.advance()
            match ch:
                case "\n":
                    self.line += 1
                case "#":
                    while self.peek() != "\n" and not(self.curr >= len(self.source)):
                        self.advance()
                case "(":
                    self.add_token(TokenType.TOK_LPAREN)
                case ")":
                    self.add_token(TokenType.TOK_RPAREN)
                case "{":
                    self.add_token(TokenType.TOK_LCURLY)
                case "}":
                    self.add_token(TokenType.TOK_RCURLY)
                case "[":
                    self.add_token(TokenType.TOK_LSQUAR)
                case "]":
                    self.add_token(TokenType.TOK_RSQUAR)
                case ",":
                    self.add_token(TokenType.TOK_COMMA)
                case ".":
                    self.add_token(TokenType.TOK_DOT)
                case "+":
                    self.add_token(TokenType.TOK_PLUS)
                case "-":
                    if self.match("-"):
                        while self.peek() != "\n" and self.curr < len(self.source):
                            self.advance()
                    else:
                        self.add_token(TokenType.TOK_MINUS)
                case "*":
                    self.add_token(TokenType.TOK_STAR)
                case "/":
                    self.add_token(TokenType.TOK_SLASH)
                case "^":
                    self.add_token(TokenType.TOK_CARET)
                case "%":
                    self.add_token(TokenType.TOK_MOD)
                case ":":
                    if self.match("="):
                        self.add_token(TokenType.TOK_ASSIGN)
                    else:
                        self.add_token(TokenType.TOK_COLON)
                case ";":
                    self.add_token(TokenType.TOK_SEMICOLON)
                case "?":
                    self.add_token(TokenType.TOK_QUESTION)
                case ">":
                    if self.match("="):
                        self.add_token(TokenType.TOK_GE)
                    elif self.match(">"):
                        self.add_token(TokenType.TOK_GTGT)
                    else:
                        self.add_token(TokenType.TOK_GT)
                case "<":
                    if self.match("="):
                        self.add_token(TokenType.TOK_LE)
                    elif self.match("<"):
                        self.add_token(TokenType.TOK_LTLT)
                    else:
                        self.add_token(TokenType.TOK_LT)
                case "=":
                    if self.match("="):
                        self.add_token(TokenType.TOK_EQ)
                case "~":
                    if self.match("="):
                        self.add_token(TokenType.TOK_NE)
                    else:
                        self.add_token(TokenType.TOK_NOT)
                case '"':
                    self.handle_string(ch)
                case "'":
                    self.handle_string(ch)
                case _:
                    if ch.isdigit():
                        self.handle_number()
                    if ch.isalpha():
                        self.handle_identifier()

File: c/lexer.c
#include "lexer.h"
#include "tokens.h"
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>

Lexer lexer_init(char *source, long source_len) {
  Lexer lexer = (Lexer){
      4, 1, 0, 0, 1, 1, source, source_len,
  };
  lexer.tokens = (Token *)malloc(sizeof(Token) * 4);
  return lexer;
}

char advance(Lexer *lexer) {
  if (lexer->curr >= lexer->source_len)
    return '\0';
  char ch = lexer->source[lexer->curr];
  lexer->curr++;
  lexer->line_position++;
  return ch;
}

void add_token(Lexer *lexer, TokenType token_type) {
  if (lexer->tokens_len == lexer->tokens_size) {
    lexer->tokens_size *= 2;
    lexer->tokens = realloc(lexer->tokens, lexer->tokens_size * sizeof(Token));
  }
  Token token =
      token_init(token_type, &lexer->source[lexer->start], lexer->line,
                 lexer->curr - lexer->start, lexer->line_position);
  lexer->tokens[lexer->tokens_len - 1] = token;
  lexer->tokens_len++;
}

char peek(Lexer *lexer) {
  if (lexer->curr >= lexer->source_len)
    return '\0';
  return lexer->source[lexer->curr];
}

char lookahead(Lexer *lexer) {
  if (lexer->curr + 1 >= lexer->source_len)
    return '\0';
  return lexer->source[lexer->curr + 1];
}

char match_char(Lexer *lexer, char expected) {
  if ((lexer->curr >= lexer->source_len) ||
      lexer->source[lexer->curr] != expected)
    return 1;
  lexer->curr++;
  return 0;
}

void handle_number(Lexer *lexer) {
  while (isdigit(peek(lexer)))
    advance(lexer);
  if ((peek(lexer) == '.') && (isdigit(lookahead(lexer)))) {
    advance(lexer);
    while (isdigit(peek(lexer)))
      advance(lexer);
    add_token(lexer, TokFloat);
  } else {
    add_token(lexer, TokInteger);
  }
}

void handle_string(Lexer *lexer, char quote) {
  while ((lexer->curr < lexer->source_len) &&
         (lexer->source[lexer->curr] != quote))
    advance(lexer);
  if (lexer->curr >= lexer->source_len) {
    printf("Unterminated string starting at line %d at position %d",
           lexer->line, lexer->line_position);
    exit(EXIT_FAILURE);
  }
  advance(lexer);
  add_token(lexer, TokString);
}

void handle_identifier(Lexer *lexer) {
  while ((isalnum(peek(lexer))) || (peek(lexer) == '_'))
    advance(lexer);
  add_token(lexer,
            keywords(&lexer->source[lexer->start], lexer->curr - lexer->start));
}

void tokenize(Lexer *lexer) {
  while (lexer->curr < lexer->source_len) {
    lexer->start = lexer->curr;
    char ch = advance(lexer);
    switch (ch) {
    case '\n':
      lexer->line++;
      lexer->line_position = 1;
      break;
    case '#':
      while ((peek(lexer) != '\n') && (lexer->curr < lexer->source_len))
        advance(lexer);
      break;
    case '(':
      add_token(lexer, TokLparen);
      break;
    case ')':
      add_token(lexer, TokRparen);
      break;
    case '{':
      add_token(lexer, TokLcurly);
      break;
    case '}':
      add_token(lexer, TokRcurly);
      break;
    case '[':
      add_token(lexer, TokLsquar);
      break;
    case ']':
      add_token(lexer, TokRsquar);
      break;
    case ',':
      add_token(lexer, TokComma);
      break;
    case '.':
      add_token(lexer, TokDot);
      break;
    case '+':
      add_token(lexer, TokPlus);
      break;
    case '-':
      if (match_char(lexer, '-') == 0) {
        while ((peek(lexer) != '\n') && (lexer->curr < lexer->source_len)) {
          advance(lexer);
        }
      } else {
        add_token(lexer, TokMinus);
      }
      break;
    case '*':
      add_token(lexer, TokStar);
      break;
    case '/':
      add_token(lexer, TokSlash);
      break;
    case '^':
      add_token(lexer, TokCaret);
      break;
    case '%':
      add_token(lexer, TokMod);
      break;
    case ':':
      if (match_char(lexer, '=') == 0) {
        add_token(lexer, TokAssign);
      } else {
        add_token(lexer, TokColon);
      }
      break;
    case ';':
      add_token(lexer, TokSemicolon);
      break;
    case '?':
      add_token(lexer, TokQuestion);
      break;
    case '>':
      if (match_char(lexer, '=') == 0) {
        add_token(lexer, TokGe);
      } else if (match_char(lexer, '>') == 0) {
        add_token(lexer, TokGtgt);
      } else {
        add_token(lexer, TokGt);
      }
      break;
    case '<':
      if (match_char(lexer, '=') == 0) {
        add_token(lexer, TokLe);
      } else if (match_char(lexer, '<') == 0) {
        add_token(lexer, TokLtlt);
      } else {
        add_token(lexer, TokLt);
      }
      break;
    case '=':
      if (match_char(lexer, '=') == 0) {
        add_token(lexer, TokEq);
      }
      break;
    case '~':
      if (match_char(lexer, '=') == 0) {
        add_token(lexer, TokNe);
      } else {
        add_token(lexer, TokNot);
      }
      break;
    case '"':
      handle_string(lexer, ch);
      break;
    case '\'':
      handle_string(lexer, ch);
      break;
    default:
      if isdigit (ch) {
        handle_number(lexer);
      } else if isalnum (ch) {
        handle_identifier(lexer);
      }
      break;
    }
  }
}
